apm-agent-core/src/main/java/co/elastic/apm/agent/impl/circuitbreaker/CircuitBreaker.java

```diff
@@ -0,0 +1,144 @@
+/*-
+ * #%L
+ * Elastic APM Java agent
+ * %%
+ * Copyright (C) 2018 - 2020 Elastic and contributors
+ * %%
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * #L%
+ */
+package co.elastic.apm.agent.impl.circuitbreaker;
+
+import co.elastic.apm.agent.context.AbstractLifecycleListener;
+import co.elastic.apm.agent.impl.ElasticApmTracer;
+import co.elastic.apm.agent.util.ExecutorUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+public class CircuitBreaker extends AbstractLifecycleListener {
+
+    private static final Logger logger = LoggerFactory.getLogger(CircuitBreaker.class);
+
+    private final ThreadPoolExecutor threadPool;
+    private final ElasticApmTracer tracer;
+    private final CircuitBreakerConfiguration circuitBreakerConfiguration;
+    private final long pollInterval;
+
+    private boolean isCurrentlyUnderStress = false;
+
+    private final List<StressMonitor> stressMonitors = new CopyOnWriteArrayList<>();
+
+    public CircuitBreaker(ElasticApmTracer tracer) {
+        this.tracer = tracer;
+        circuitBreakerConfiguration = tracer.getConfig(CircuitBreakerConfiguration.class);
+        pollInterval = circuitBreakerConfiguration.getStressMonitoringPollingInterval();
+        threadPool = ExecutorUtils.createSingleThreadDeamonPool("circuit-breaker", 1);
+    }
+
+    @Override
+    public void start(ElasticApmTracer tracer) {
+        // failsafe loading of stress monitors in isolation
+        loadGCStressMonitor(tracer);
+
+        threadPool.execute(new Runnable() {
+            @Override
+            public void run() {
+                pollStressMonitors();
+            }
+        });
+    }
+
+    private void loadGCStressMonitor(ElasticApmTracer tracer) {
+        try {
+            stressMonitors.add(new GCStressMonitor(tracer));
+        } catch (Throwable throwable) {
+            logger.error("Failed to load the GC stress monitor. Circuit breaker will not be triggered based on GC events.", throwable);
```

Comment:
what kind of failures do we expect here ? shouldn't they be managed at monitor implementation level ?
if there are missing JMX attributes for example, we should rather have a monitor that does never calls pause/resume than an exception thrown from constructor.

Reply:
> what kind of failures do we expect here ? 

Not expecting, just watching out from

> shouldn't they be managed at monitor implementation level ?

Yes, unless the class itself fails to load, eg because of missing dependencies. Since it is JMX, I was cautious even though we ended up using a generic API. I can see why it seems overly cautious, but if there's no reason to remove, let's leave it.

> if there are missing JMX attributes for example, we should rather have a monitor that does never calls pause/resume than an exception thrown from constructor.

I thought more about failure to load the class, rather than an exception in the constructor. This guard is not about the concern of pausing/resuming, it is in order to make sure we do not fail to register the next monitor if this one cannot be loaded.